Add GitHub Actions

outofambit opened the pull request:
_[...]_

outofambit committed:
first try

outofambit committed:
maybe we don't need this

outofambit committed:
checkout with submodules recursively

shiftkey commented:
@outofambit I noticed a similar issue on Azure Pipelines recently, and added https://github.com/shiftkey/desktop/commit/42f219c35bb2890236622eaff5778271e1129af3 to workaround the issue

outofambit committed:
mock nativeTheme for tests

outofambit commented on app/test/__mocks__/electron.ts:
```diff
@@ -15,6 +15,11 @@ export const remote = {
     on: jest.fn(),
   },
+  nativeTheme: {
+    addListeners: jest.fn(),
```
```suggestion
    addListener: jest.fn(),
```

VChernomYA replied:
-    addListeners: jest.fn(),
+    addListener: jest.fn(),

niik committed:
Merge branch 'development' into spike/actions-win32

niik committed:
Set CI env var for all steps

See https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables

niik committed:
Get SHA on GitHub Actions

niik committed:
Get whether we're on a fork in an action

See https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables

niik committed:
We can build on Windows on GH actions as well

niik committed:
This is already set by default

See https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables

niik committed:
GitHub Actions can build on darwin as well

niik committed:
Skip this for now

We don't use listWorkTrees and it's causing problems on GitHub actions

niik committed:
We're on Node 12 now

niik committed:
More apt description

niik committed:
Cache yarn packages

niik committed:
Don't know how I managed to duplicate that

niik committed:
Try bumping spectron?

niik committed:
Limit the run time

niik committed:
Use the correct version of spectron

niik committed:
Build it

niik committed:
Lint it

niik committed:
Give it a little bit more time

niik committed:
Try inspecting what's changed

niik committed:
Oh wait, I'm at fault, as per usual

niik committed:
Add some secrets

niik committed:
Add an initial publish step (don't worry, it doesn't have the creds to publish yet)

niik committed:
Split up unit and integration tests

Temporarily so that we can debug integration tests on their own

niik committed:
How hard could it possibly be?

niik committed:
Maybe it's time to stop referring to this as win32 only

niik committed:
Rename action to ci

niik committed:
Whoops

niik committed:
Try a nicer job name

niik committed:
Test reporting test errors

niik committed:
I guess not?

niik committed:
Include test locations

niik committed:
Drop jest-junit

niik committed:
Revert accidental changes to jest unit config

niik committed:
Include test location in results for integration tests

niik committed:
We don't have jest-unit any more

niik committed:
Debug forever

niik committed:
omg

niik committed:
Get rid of debug log

niik committed:
Try getting the actual failure position

niik committed:
Simpler path

niik committed:
Try splitting it out even further to see if we can get windows builds to fail

niik committed:
Try forcing an eslint error

rafeca committed:
Revert "Try forcing an eslint error"

This reverts commit 58fb28005b56357ce7fe11f56693de4217ac4ecb.

rafeca committed:
Remove test that was failing on purpose

rafeca committed:
Run the GitHub Actions workflow on release branches

rafeca committed:
Add missing secrets

rafeca commented:
I've started seeing build failures in the `yarn check-modified` step on GitHub Actions under Windows, it seems like the `app/static/common/gitignore` is causing a dirty tree due to the symlinks that it has:

After checkout, this is what we got:

```
$ git diff
diff --git a/app/static/common/gitignore b/app/static/common/gitignore
--- a/app/static/common/gitignore
+++ b/app/static/common/gitignore
@@ -1 +1 @@
-Subproject commit 2644536a24f37d3906c7ea56e57f6965b1e21f36
+Subproject commit 2644536a24f37d3906c7ea56e57f6965b1e21f36-dirty

$ cd app/static/common/gitignore && git status
HEAD detached at 2644536
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   Clojure.gitignore
	modified:   Fortran.gitignore
	modified:   Global/Octave.gitignore
	modified:   Kotlin.gitignore

no changes added to commit (use "git add" and/or "git commit -a")
```

Note that the modified files correspond to the symlinks in the original repo: https://github.com/github/gitignore

Link to the Run: https://github.com/desktop/desktop/runs/897975189#step:7:40

I'm investigating why this didn't fail before (https://github.com/desktop/desktop/runs/854464821) but now it does... 🤔

rafeca commented:
Ok, looks like the Windows image has been changed and it's quite possible that an upgraded version of `git` or another dependency is causing this change.

* In 48fe311 (a build that passed) the virtual environment [was v20200630.0](https://github.com/desktop/desktop/runs/854464821#step:1:8) ([list of installed software](https://github.com/actions/virtual-environments/blob/win19/20200630.0/images/win/Windows2019-Readme.md)).
* In 00d8c32 (a build that failed) the virtual environment [was 20200714.1](https://github.com/desktop/desktop/runs/897975189#step:1:8) ([list of installed software](https://github.com/actions/virtual-environments/blob/win19/20200714.1/images/win/Windows2019-Readme.md)).

These are the changes between the two virtual environment versions: https://github.com/actions/virtual-environments/commit/f7c1a0264ff7112855c9c19c48dfbdb42526a7c1 and https://github.com/actions/virtual-environments/commit/95222e07e94ec86d400c670bdc2f92f5a2b20fd0 .

The only suspecting thing that I see is "Enabled Windows Subsystem for Linux", but even this seems unlikely to cause this problem, I'm going to keep investigating.

Anyways, I think that in order to keep our builds stable and consistent, we should pin to specific VMs and not use latest, going to investigate this as well.

rafeca commented:
To verify that the failure is caused by the change of the VM image that GitHub Actions is using I tried re-running a job on an old commit that used to have CI passing (f4aace4).

* Old run that passed with the previous VM: https://github.com/desktop/desktop/runs/854271314
* New run that failed with the current VM: https://github.com/desktop/desktop/runs/898140396

I cannot find any information about pinning our CI runs to a specific version of the VM (in the [official docs](https://docs.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#ip-addresses-of-github-hosted-runners) it mentions using either `windows-latest` or `windows-2019`, implying that both strings will use the latest available VM for Windows 2019.

rafeca committed:
Pin to major versions for OS images

rafeca commented:
Oh god, this was a tricky one to debug! turns out that `core.symlinks` was enabled by default in the VMs recently (see https://github.com/actions/virtual-environments/pull/1186). Since there was no reference to this in the VM release changelogs I couldn't find this by myself.

Long story short, it looks like Git for Windows has some issues when `core.symlinks` is enabled on repositories with submodules. I'm going to create an issue on that repository.

rafeca committed:
Disable core.symlinks on windows

rafeca commented:
Ok, I've added a few commits that hopefully will make this PR green again:

- d57b390 and 033cb0c revert the failures that were added in purpose to test that CI was reporting errors correctly.
- ce6228b enables GitHub Actions on pushes to release branches.
- 9398db9 adds the missing secrets as environment variables for the publish step.
- 853dfea pins the VM version (only to the major version of the VM but that's better than nothing).
- 1d6b7e9 disables symlink support on the Windows runs (this is equivalent to what happened in AppVeyor).

With this, I hope that the build turns green 🤞

Two things to take into account:

- Would be good to enable symlinks support on Windows. Since we don't have it enabled there's a bug in Windows where selecting certain `gitignore` templates (like Kotlin) when creating a new repo in Windows results in wrong contents of the `.gitignore` file. This is a long-standing bug and I don't think we can fix it properly without a fix in Git for Windows.
- Would also be good to pin to a specific VM version to have more stable builds in CI. Unfortunately, after talking to the GitHub Actions team this is only possible when using our own self-hosted VMs for CI. We don't want to go through that path (at least for now) so we'll have to live with VMs that self-update automatically.

rafeca committed:
Calculate the release branch name correctly on GitHub Actions

rafeca committed:
Add some secrets to the build step

Some secrets are needed in the build step for macOS

rafeca committed:
Return SHA for github actions

rafeca committed:
Guard against empty value env variables in GitHub Actions

rafeca commented:
I've added a few more commits to fix publishing releases. With this, I've been able to release successfully `v2.5.4-test3`.

Some things I'd like to improve from the CI setup:

1. Run the tests before publishing the release: We currently run the tests afterwards, which doesn't make a lot of sense 😅 
2. Only run the `yarn publish` step on releases branch: we currently run it always and the code decides whether to do something or not.
3. Only pass the secrets when building a release: we currently always pass the secrets, so it would be better to have a dedicated workflow for releasing and only pass the secrets in that workflow.

rafeca committed:
Execute tests before publishing the production app

rafeca committed:
Remove skip from test

It's better to test the head than to not test anything.

Unfortunately it's quite complex to fix the test. The problem comes from the `temp` NPM package, which uses the `TEMP` env variable in Windows machines.

For compatibility reasons, the `TEMP` env var in Windows returns the shortland DOS 8.3 version of the temp folder, which means that folders/files with more than 8 characters get a different name. In GitHub Actions the username is `runneradmin` which gets translated in DOS 8.3 to `RUNNER~1`.

The ideal solution would be to translate the shortland names back their full name be on the `setupEmptyRepository()` but it's quite tricky to do that (we would need to spawn `cmd` and do nasty internal windows calls).

Since this is the only test checking a path I think it's more pragmatic to just remove the path check.

rafeca commented:
> Long story short, it looks like Git for Windows has some issues when core.symlinks is enabled on repositories with submodules. I've created a small repository that reproduces these issues (https://github.com/rafeca/test-submodules-with-symlinks) and I'm planning to create an issue in the git-for-windows repository.

I've checked with the newly released v2.28.0 version of git for windows and these issues are fixed :tada:, this means that once the GitHub Actions VMs get updated with this new git version (which is happening very very soon: https://github.com/actions/virtual-environments/pull/1349 😃 ), we'll be able to revert 1d6b7e90483ee7c2548516116fae74a825ab670e.

niik committed:
Merge branch 'development' into spike/actions-win32

niik committed:
Resync yarn.lock after merge

niik committed:
Merge branch 'development' into spike/actions-win32

niik committed:
Resync yarn lock after  merge

niik committed:
Remove appveyor and travis configs

rafeca committed:
Revert "Disable core.symlinks on windows"

This reverts commit 1d6b7e90483ee7c2548516116fae74a825ab670e.

rafeca committed:
Merge branch 'development' into spike/actions-win32

rafeca commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,77 @@
+  push:
+    branches:
+      - development
+      - releases/*
```
I don't think we need to explicitly build the `releases/*` branches, I added this while testing...

```suggestion
```

niik replied:
Yeah, that makes sense to me

rafeca reviewed (dismissed):
LGTM

niik committed:
Update .github/workflows/ci.yml

Co-authored-by: Rafael Oleza <rafeca@github.com>

niik committed:
This is how I would like it to work

niik committed:
is rafael cleverer than a fox?

niik committed:
INDENTATION THOUGH

rafeca committed:
Remove branches filtering from pull_requests event

I don't know if this is causing PRs from forks to not run...

niik committed:
Use exact cache match if it exists

niik committed:
Just the OS  is enough I think

rafeca reviewed (approved):
LGTM! :shipit:

outofambit commented:
OH EM GEE

<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia2.giphy.com%2Fmedia%2FL0ZuUuWbOScpA3No8f%2Fgiphy.gif&token=oh_em_gee" data-gitmeme-token="oh_em_gee"><img src="https://media2.giphy.com/media/L0ZuUuWbOScpA3No8f/giphy.gif" title="Created by gitme.me with /oh_em_gee"/></a>

SecretRose7 reviewed (commented):
FiireFox7_RoseHub Enterprise software systems are the best way to go
